gio/src/unix_fd_list.rs

```diff
@@ -42,39 +42,43 @@ pub trait UnixFDListExtManual: IsA<UnixFDList> + Sized {
     }
 
     #[doc(alias = "g_unix_fd_list_get")]
-    fn get(&self, index_: i32) -> Result<RawFd, glib::Error> {
+    fn get(&self, index_: i32) -> Result<OwnedFd, glib::Error> {
         unsafe {
             let mut error = ptr::null_mut();
-            let ret = ffi::g_unix_fd_list_get(self.as_ref().to_glib_none().0, index_, &mut error);
+            let raw_fd =
+                ffi::g_unix_fd_list_get(self.as_ref().to_glib_none().0, index_, &mut error);
+            let fd = OwnedFd::from_raw_fd(raw_fd);
             if error.is_null() {
-                Ok(ret)
+                Ok(fd)
             } else {
                 Err(from_glib_full(error))
             }
         }
     }
 
     #[doc(alias = "g_unix_fd_list_peek_fds")]
-
-    fn peek_fds(&self) -> Vec<RawFd> {
+    fn peek_fds(&self) -> Vec<BorrowedFd<'_>> {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_none_num(
+            let ret: Vec<RawFd> = FromGlibContainer::from_glib_none_num(
                 ffi::g_unix_fd_list_peek_fds(self.as_ref().to_glib_none().0, length.as_mut_ptr()),
                 length.assume_init() as usize,
             );
-            ret
+            ret.into_iter()
+                .map(|fd| BorrowedFd::borrow_raw(fd))
+                .collect()
         }
     }
+
     #[doc(alias = "g_unix_fd_list_steal_fds")]
-    fn steal_fds(&self) -> Vec<RawFd> {
+    fn steal_fds(&self) -> Vec<OwnedFd> {
         unsafe {
             let mut length = mem::MaybeUninit::uninit();
-            let ret = FromGlibContainer::from_glib_full_num(
+            let ret: Vec<RawFd> = FromGlibContainer::from_glib_full_num(
```

Comment:
Done, but the return type is still `Vec<BorrowedFd<'_>>`, which is kinda odd.

Reply:
It says `steal` here so they're probably not borrowed but owned?

Also why don't you use `glib::Slice` here to avoid the copy to a `Vec`?

Reply:
I wasn't able to do the map RawFd->BorrowedFd without ownership issues at the time.

Reply:
This can actually just return `&[BorrowedFd]`. You can use `std::slice::from_raw_parts()` plus a cast. `BorrowedFd` is guaranteed to have the same layout as an actual fd.

Reply:
`OwnedFd` has the same layout as an actual fd so you could just cast things.

I think the correct thing to do here (because `glib::Slice` requires all the traits for the contained type) is to define a new `struct FdArray(...)` that stores a `glib::Slice<i32>` and can `Deref` into `BorrowedFd` and `IntoIterator<Item = OwnedFd>`, and on `Drop` frees things accordingly.